File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/DiaryYearMonthListItem.java
package com.websarva.wings.android.zuboradiary.ui.list;

import android.os.Build;

import androidx.annotation.NonNull;

import java.time.Year;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class DiaryYearMonthListItem implements Cloneable {
    private final String id = UUID.randomUUID().toString();
    private YearMonth yearMonth;
    private List<DiaryDayListItem> diaryDayListItemList;
    private int viewType;

    public DiaryYearMonthListItem(int viewType) {
        yearMonth = null;
        diaryDayListItemList = new ArrayList<>();
        this.viewType = viewType;
    }

    public DiaryYearMonthListItem(
            @NonNull YearMonth yearMonth,
            @NonNull List<DiaryDayListItem> diaryDayListItemList, int viewType) {
        this.yearMonth = yearMonth;
        this.diaryDayListItemList = diaryDayListItemList;
        this.viewType = viewType;
    }

    // Object#clone例外処理:https://yujisoftware.hatenablog.com/entry/CloneNotSupportedException
    @NonNull
    @Override
    public DiaryYearMonthListItem clone() {
        DiaryYearMonthListItem clone = null;
        try {
            clone = (DiaryYearMonthListItem) super.clone();
            List<DiaryDayListItem> cloneDiaryDayList = new ArrayList<>();
            for (DiaryDayListItem item : this.diaryDayListItemList) {
                DiaryDayListItem cloneItem = item.clone();
                cloneDiaryDayList.add(cloneItem);
            }
            clone.diaryDayListItemList = cloneDiaryDayList;
        } catch (CloneNotSupportedException e) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                throw new InternalError(e);
            } else {
                throw new RuntimeException(e.getMessage());
            }
        }
        return clone;
    }

    public String getId() {
        return id;
    }

    public YearMonth getYearMonth() {
        return this.yearMonth;
    }

    public void setYearMonth(YearMonth yearMonth) {
        this.yearMonth = yearMonth;
    }

    public List<DiaryDayListItem> getDiaryDayListItemList() {
        return diaryDayListItemList;
    }

    public void setDiaryDayListItemList(List<DiaryDayListItem> diaryDayListItemList) {
        this.diaryDayListItemList = diaryDayListItemList;
    }

    public int getViewType() {
        return viewType;
    }

    public void setViewType(int viewType) {
        this.viewType = viewType;
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/domain/model/list/diary/DiaryYearMonthListItem.kt
package com.websarva.wings.android.zuboradiary.domain.model.list.diary

import java.time.YearMonth

/**
 * 年月ごとの日記リストの各アイテムを表す基底クラス。
 *
 * このクラスは、年月単位でグループ化された日記リストの要素の共通の型を提供する。
 * 具体的なアイテムの種類（日記の年月ごとの集まり、日記なしメッセージ、プログレスインジケータ）に応じて、
 * このクラスを継承したサブクラスが定義される。
 *
 * @param T [DiaryDayListItem] を実装するアイテムの型。これは [Diary] サブクラス内の日記リストの型として使用される。
 */
internal sealed class DiaryYearMonthListItem<T : DiaryDayListItem> {

    /**
     * 特定の年月に属する日記のリストを表すアイテム。
     *
     * @param T [DiaryDayListItem] を実装するアイテムの型。
     * @property yearMonth この日記リストが対象とする年月。
     * @property diaryDayList その年月の対象となる日単位の日記リスト。
     */
    data class Diary<T : DiaryDayListItem>(
        val yearMonth: YearMonth,
        val diaryDayList: DiaryDayList<T>
    ) : DiaryYearMonthListItem<T>()

    /**
     * 日記が存在しないことを示すメッセージアイテム。
     *
     * @param T [DiaryDayListItem] を実装するアイテムの型。
     */
    class NoDiaryMessage<T : DiaryDayListItem> : DiaryYearMonthListItem<T>()

    /**
     * 続きの日記リスト読み込み中を示すプログレスインジケータアイテム。
     *
     * @param T [DiaryDayListItem] を実装するアイテムの型。
     */
    class ProgressIndicator<T : DiaryDayListItem> : DiaryYearMonthListItem<T>()
}
